File: crates/cogit-core/src/store.rs
//! Content-addressed object store: zlib(<type> <size>\0<canonical-json>).
//! Write: validate -> canonicalize -> hash -> tmp file -> atomic rename.
//! Read: decompress -> header -> size, hash, canonical-bytes, schema checks.

use std::fs;
use std::io::{Read, Write};
use std::path::{Path, PathBuf};

use flate2::read::ZlibDecoder;
use flate2::write::ZlibEncoder;
use flate2::Compression;
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::canonical::{canonical_json, parse_json};
use crate::error::{CoreError, Result};
use crate::objects::{encode_object, is_oid, validate_object, OBJECT_TYPES};

pub struct ObjectStore {
    objects_dir: PathBuf,
    tmp_dir: PathBuf,
}

impl ObjectStore {
    pub fn new(cogit_dir: &Path) -> Self {
        ObjectStore {
            objects_dir: cogit_dir.join("objects"),
            tmp_dir: cogit_dir.join("tmp"),
        }
    }

    pub fn path_for(&self, oid: &str) -> Result<PathBuf> {
        if !is_oid(oid) {
            return Err(CoreError::User(format!("object store: invalid object id '{oid}'")));
        }
        let hexpart = &oid[7..];
        Ok(self.objects_dir.join(&hexpart[..2]).join(&hexpart[2..]))
    }

    pub fn exists(&self, oid: &str) -> Result<bool> {
        Ok(self.path_for(oid)?.is_file())
    }

    /// Write an object; deduplicates by hash. Returns the object ID.
    pub fn write(&self, value: &Value) -> Result<String> {
        let (oid, preimage) = encode_object(value)?;
        let target = self.path_for(&oid)?;
        if target.exists() {
            let existing = self.read_preimage(&oid, &target)?;
            if existing != preimage {
                return Err(CoreError::Corruption(format!(
                    "object store: {oid} exists with different content (collision/corruption)"
                )));
            }
            return Ok(oid);
        }
        fs::create_dir_all(target.parent().expect("fanout parent"))?;
        fs::create_dir_all(&self.tmp_dir)?;

        let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
        encoder.write_all(&preimage)?;
        let compressed = encoder.finish()?;

        let tmp_path = self.tmp_dir.join(format!(
            "obj-{}-{}",
            std::process::id(),
            &oid[7..19],
        ));
        {
            let mut file = fs::File::create(&tmp_path)?;
            file.write_all(&compressed)?;
            file.sync_all()?;
        }
        fs::rename(&tmp_path, &target)?;
        Ok(oid)
    }

    /// Read and fully verify an object.
    pub fn read(&self, oid: &str) -> Result<Value> {
        let target = self.path_for(oid)?;
        if !target.is_file() {
            return Err(CoreError::User(format!("object store: {oid} not found")));
        }
        let preimage = self.read_preimage(oid, &target)?;
        self.decode_preimage(oid, &preimage)
    }

    fn read_preimage(&self, oid: &str, path: &Path) -> Result<Vec<u8>> {
        let compressed = fs::read(path)?;
        let mut decoder = ZlibDecoder::new(compressed.as_slice());
        let mut preimage = Vec::new();
        decoder
            .read_to_end(&mut preimage)
            .map_err(|e| CoreError::Corruption(format!("object store: {oid} corrupt zlib body: {e}")))?;
        Ok(preimage)
    }

    fn decode_preimage(&self, oid: &str, preimage: &[u8]) -> Result<Value> {
        let nul = preimage
            .iter()
            .position(|b| *b == 0)
            .ok_or_else(|| CoreError::Corruption(format!("object store: {oid} malformed header (no NUL)")))?;
        let header = std::str::from_utf8(&preimage[..nul])
            .map_err(|_| CoreError::Corruption(format!("object store: {oid} malformed header")))?;
        let body = &preimage[nul + 1..];
        let (type_text, size_text) = header
            .split_once(' ')
            .ok_or_else(|| CoreError::Corruption(format!("object store: {oid} malformed header")))?;
        let declared_size: usize = size_text
            .parse()
            .map_err(|_| CoreError::Corruption(format!("object store: {oid} malformed header")))?;
        if !OBJECT_TYPES.contains(&type_text) {
            return Err(CoreError::Corruption(format!(
                "object store: {oid} unknown object type '{type_text}'"
            )));
        }
        if declared_size != body.len() {
            return Err(CoreError::Corruption(format!(
                "object store: {oid} size mismatch (declared {declared_size}, actual {})",
                body.len()
            )));
        }
        let digest = Sha256::digest(preimage);
        let mut computed = String::with_capacity(71);
        computed.push_str("sha256:");
        for byte in digest {
            computed.push_str(&format!("{byte:02x}"));
        }
        if computed != oid {
            return Err(CoreError::Corruption(format!(
                "object store: hash-path mismatch (path {oid}, content {computed})"
            )));
        }
        let body_text = std::str::from_utf8(body)
            .map_err(|_| CoreError::Corruption(format!("object store: {oid} invalid JSON body")))?;
        let value = parse_json(body_text)
            .map_err(|e| CoreError::Corruption(format!("object store: {oid} invalid JSON body: {e}")))?;
        let obj_type = validate_object(&value)
            .map_err(|e| CoreError::Corruption(format!("object store: {oid} schema invalid: {e}")))?;
        if canonical_json(&value)?.as_bytes() != body {
            return Err(CoreError::Corruption(format!(
                "object store: {oid} body is not canonical JSON"
            )));
        }
        if obj_type != type_text {
            return Err(CoreError::Corruption(format!(
                "object store: {oid} header/body type mismatch"
            )));
        }
        Ok(value)
    }
}

File: prototype/cogit/store.py
"""Content-addressed object store: zlib(<type> <size>\\0<canonical-json>).

Write path: validate -> canonicalize -> hash -> tmp file -> atomic rename.
Read path: decompress -> parse header -> verify size, hash, canonical bytes,
schema. Contract: docs/spec/object-format-v1.md.
"""

import hashlib
import os
import re
import zlib

from .canonical import canonical_json, parse_json
from .errors import CorruptionError, UserError
from .objects import OBJECT_TYPES, encode_object, is_oid, validate_object


class ObjectStore:
    def __init__(self, cogit_dir):
        self.objects_dir = os.path.join(cogit_dir, "objects")
        self.tmp_dir = os.path.join(cogit_dir, "tmp")

    def path_for(self, oid: str) -> str:
        if not is_oid(oid):
            raise UserError(f"object store: invalid object id '{oid}'")
        hexpart = oid.split(":", 1)[1]
        return os.path.join(self.objects_dir, hexpart[:2], hexpart[2:])

    def exists(self, oid: str) -> bool:
        return os.path.isfile(self.path_for(oid))

    PREFIX_RE = re.compile(r"^[0-9a-f]{6,64}$")

    def expand_prefix(self, name: str) -> str:
        """Expand a unique object-id prefix (>= 6 hex chars) to a full oid (COG-025)."""
        hexpart = name[len("sha256:") :] if name.startswith("sha256:") else name
        if not self.PREFIX_RE.match(hexpart):
            raise UserError(
                f"object store: '{name}' is not an object id or unique prefix (>= 6 hex chars)"
            )
        if len(hexpart) == 64:
            return "sha256:" + hexpart
        fanout_dir = os.path.join(self.objects_dir, hexpart[:2])
        rest = hexpart[2:]
        matches = []
        if os.path.isdir(fanout_dir):
            matches = [entry for entry in sorted(os.listdir(fanout_dir)) if entry.startswith(rest)]
        if not matches:
            raise UserError(f"object store: no object matches prefix '{name}'")
        if len(matches) > 1:
            raise UserError(f"object store: prefix '{name}' is ambiguous ({len(matches)} matches)")
        return f"sha256:{hexpart[:2]}{matches[0]}"

    def write(self, obj) -> str:
        """Write an object; deduplicates by hash. Returns the object ID."""
        oid, preimage = encode_object(obj)
        target = self.path_for(oid)
        if os.path.exists(target):
            # Same path must mean same content; anything else is corruption.
            existing = self._read_preimage(oid, target)
            if existing != preimage:
                raise CorruptionError(
                    f"object store: {oid} exists with different content (collision/corruption)"
                )
            return oid
        os.makedirs(os.path.dirname(target), exist_ok=True)
        os.makedirs(self.tmp_dir, exist_ok=True)
        compressed = zlib.compress(preimage)
        fd, tmp_path = self._mkstemp()
        try:
            with os.fdopen(fd, "wb") as handle:
                handle.write(compressed)
                handle.flush()
                os.fsync(handle.fileno())
            os.rename(tmp_path, target)
        finally:
            if os.path.exists(tmp_path):
                os.unlink(tmp_path)
        return oid

    def read(self, oid: str) -> dict:
        """Read and fully verify an object."""
        target = self.path_for(oid)
        if not os.path.isfile(target):
            raise UserError(f"object store: {oid} not found")
        preimage = self._read_preimage(oid, target)
        return self._decode_preimage(oid, preimage)

    def _mkstemp(self):
        import tempfile

        return tempfile.mkstemp(prefix="obj-", dir=self.tmp_dir)

    def _read_preimage(self, oid: str, path: str) -> bytes:
        with open(path, "rb") as handle:
            compressed = handle.read()
        try:
            return zlib.decompress(compressed)
        except zlib.error as exc:
            raise CorruptionError(f"object store: {oid} corrupt zlib body: {exc}") from exc

    def _decode_preimage(self, oid: str, preimage: bytes) -> dict:
        nul = preimage.find(b"\x00")
        if nul < 0:
            raise CorruptionError(f"object store: {oid} malformed header (no NUL)")
        header = preimage[:nul]
        body = preimage[nul + 1 :]
        try:
            type_text, size_text = header.decode("ascii").split(" ", 1)
            declared_size = int(size_text)
        except (UnicodeDecodeError, ValueError) as exc:
            raise CorruptionError(f"object store: {oid} malformed header") from exc
        if type_text not in OBJECT_TYPES:
            raise CorruptionError(f"object store: {oid} unknown object type '{type_text}'")
        if declared_size != len(body):
            raise CorruptionError(
                f"object store: {oid} size mismatch (declared {declared_size}, actual {len(body)})"
            )
        computed = "sha256:" + hashlib.sha256(preimage).hexdigest()
        if computed != oid:
            raise CorruptionError(f"object store: hash-path mismatch (path {oid}, content {computed})")
        try:
            obj = parse_json(body.decode("utf-8"))
        except UserError:
            raise
        except (UnicodeDecodeError, ValueError) as exc:
            raise CorruptionError(f"object store: {oid} invalid JSON body") from exc
        try:
            # ADR-0015: reads tolerate unknown fields (forward compatibility);
            # writes stay strict — see ObjectStore.write
            validate_object(obj, mode="read")
        except UserError as exc:
            raise CorruptionError(f"object store: {oid} schema invalid: {exc}") from exc
        if canonical_json(obj).encode("utf-8") != body:
            raise CorruptionError(f"object store: {oid} body is not canonical JSON")
        if obj["type"] != type_text:
            raise CorruptionError(f"object store: {oid} header/body type mismatch")
        return obj
